File: PA_55.py
#LC 102 - Binary Tree Level Order Traversal
#Time Complexity - O(n)
#Space Complexity - O(h)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        self.ans = []
        def helper(root,depth):
            if not root:
                return
            if depth == len(self.ans):
                self.ans.append([])
            self.ans[depth].append(root.val)
            helper(root.left,depth+1)
            helper(root.right,depth+1)
        helper(root,0)
        return(self.ans)
        

File: PA55.java
//Leetcode 102. Binary Tree Level Order Traversal
//BFS solution
//Time Complexity : O(n)  // n is the number of nodes in the tree
//Space Complexity: O(n) //size of Queue, where size is equal to max no nodes at a level in tree i.e n/2
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if(root==null) return res;
        
            Queue<TreeNode> q= new LinkedList<>();
            q.offer(root);
            while(!q.isEmpty()){
                int s=q.size();
                List<Integer> ls = new ArrayList<>();
                for(int i=0; i<s;i++){
                    TreeNode t= q.poll();
                    ls.add(t.val);
                    if(t.left!=null) q.offer(t.left);
                    if(t.right!=null) q.offer(t.right);
                }
                res.add(ls);
            }
        return res;
    }
}


======================================================================
//Leetcode 102. Binary Tree Level Order Traversal
//DFS solution:
//Time Complexity : O(n)  // n is the number of nodes in the tree
//Space Complexity: O(n) // size of Stack, where size is equal to height of the tree i.e log(n)
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> levelOrder(TreeNode root) {
        res = new ArrayList<List<Integer>>();
        if(root==null) return res;                   
        dfs(root,0);
        return res;       
    } 
    private void dfs(TreeNode root, int level){
        if(root==null){
            return ;
        }        
        if(res.size()==level){
            res.add(new ArrayList<Integer>());
        }
        res.get(level).add(root.val);      
        dfs(root.left,level+1);
        dfs(root.right,level+1);
        
    }
}
==========================================================================
